From 521967559668f1e02b9adc69ee6d5f5b15aacce4 Mon Sep 17 00:00:00 2001
From: Hugues Fruchet <hugues.fruchet@foss.st.com>
Date: Wed, 31 Jan 2024 16:54:56 +0100
Subject: [PATCH 0884/1141] hantro: h1: workaround for pink artefacts on P
 frames

It seems that hardware expects a negative value on 16bits
for inter_favor otherwise P frames have pink artefacts.

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>
Change-Id: I7d8f01703073ceccd40d9f30af3cad796f9807b0
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/362857
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Hugues FRUCHET <hugues.fruchet@st.com>
Tested-by: Hugues FRUCHET <hugues.fruchet@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
---
 drivers/media/platform/verisilicon/hantro_h1_vp8_enc.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

--- a/drivers/media/platform/verisilicon/hantro_h1_vp8_enc.c
+++ b/drivers/media/platform/verisilicon/hantro_h1_vp8_enc.c
@@ -1208,7 +1208,7 @@ static void hantro_h1_vp8_enc_set_params
 	u32 deq;
 	u32 tmp;
 	u32 reg;
-	s32 inter_favor;
+	s16 inter_favor;
 
 	vepu_write_relaxed(vpu, 0, H1_REG_STR_HDR_REM_MSB);
 	vepu_write_relaxed(vpu, 0, H1_REG_STR_HDR_REM_LSB);
@@ -1220,7 +1220,12 @@ static void hantro_h1_vp8_enc_set_params
 
 	inter_favor = 128 - ctx->vp8_enc.prob_intra;
 	if (inter_favor >= 0)
-		inter_favor = max(0, (int32_t)(qp * 2 - 40));
+		/*
+		 * It seems that hardware expects a negative s16 non-0 value
+		 * for inter_favor otherwise P frames have pink artefacts...
+		 */
+		/*inter_favor = max(0, (int32_t)(qp * 2 - 40));*/
+		inter_favor = -max(1, (int32_t)(qp * 2 - 40));
 
 	reg |= H1_REG_ENC_CTRL2_INTRA16X16_MODE(qp * 1024 / 128);
 	vepu_write_relaxed(vpu, reg, H1_REG_ENC_CTRL2);//in VEPU_REG_INTRA_INTER_MODE
@@ -1515,7 +1520,6 @@ int hantro_h1_vp8_enc_run(struct hantro_
 	/* Start the hardware. */
 	reg =     H1_REG_ENC_CTRL_TIMEOUT_EN
 		| H1_REG_ENC_CTRL_MV_WRITE //for VEPU_REG_MV_WRITE_EN in VEPU_REG_INTERRUPT
-		| H1_REG_ENC_REC_WRITE_DISABLE//FIXME needed to have visually correct P frames
 		| H1_REG_ENC_CTRL_WIDTH(MB_WIDTH(ctx->src_fmt.width))
 		| H1_REG_ENC_CTRL_HEIGHT(MB_HEIGHT(ctx->src_fmt.height))
 		| H1_REG_ENC_CTRL_ENC_MODE_VP8
